cmd/snap-repair/runner.go

```diff
@@ -185,3 +211,268 @@ func (run *Runner) Peek(brandID, repairID string) (headers map[string]interface{
 
 	return headers, nil
 }
+
+// deviceInfo captures information about the device.
+type deviceInfo struct {
+	Brand string `json:"brand"`
+	Model string `json:"model"`
+}
+
+// RepairStatus represents the possible statuses of a repair.
+type RepairStatus int
+
+const (
+	RetryStatus RepairStatus = iota
+	NotApplicableStatus
+	DoneStatus
+)
+
+// RepairState holds the current revision and status of a repair in a sequence of repairs.
+type RepairState struct {
+	Seq      int          `json:"seq"`
+	Revision int          `json:"revision"`
+	Status   RepairStatus `json:"status"`
+}
+
+// state holds the atomically updated control state of the runner with sequences of repairs and their states.
+type state struct {
+	Device    deviceInfo                `json:"device"`
+	Sequences map[string][]*RepairState `json:"sequences,omitempty"`
+}
+
+func (run *Runner) readState() error {
+	r, err := os.Open(dirs.SnapRepairStateFile)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	dec := json.NewDecoder(r)
+	return dec.Decode(&run.state)
+}
+
+func (run *Runner) initState() error {
+	if err := os.MkdirAll(dirs.SnapRepairDir, 0775); err != nil {
+		panic(fmt.Sprintf("cannot create repair state directory: %v", err))
+	}
+	// best-effort remove old
+	os.Remove(dirs.SnapRepairStateFile)
+	// TODO: init Device
+	run.SaveState()
+	return nil
+}
+
+// LoadState loads the repairs' state from disk, and (re)initializes it if it's missing or corrupted.
+func (run *Runner) LoadState() error {
+	err := run.readState()
+	if err == nil {
+		return nil
+	}
+	// error => initialize from scratch
+	// TODO: log error?
+	return run.initState()
+}
+
+// SaveState saves the repairs' state to disk.
+func (run *Runner) SaveState() {
+	m, err := json.Marshal(&run.state)
+	if err != nil {
+		panic(fmt.Sprintf("cannot marshal repair state: %v", err))
+	}
+	err = osutil.AtomicWriteFile(dirs.SnapRepairStateFile, m, 0600, 0)
+	if err != nil {
+		panic(fmt.Sprintf("cannot save repair state: %v", err))
+	}
+}
+
+func stringList(headers map[string]interface{}, name string) ([]string, error) {
+	v, ok := headers[name]
+	if !ok {
+		return nil, nil
+	}
+	l, ok := v.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("invalid %q header", name)
+	}
+	r := make([]string, len(l))
+	for i, v := range l {
+		s, ok := v.(string)
+		if !ok {
+			return nil, fmt.Errorf("invalid %q header", name)
+		}
+		r[i] = s
+	}
+	return r, nil
+}
+
+// Applicable returns whether a repair with the given headers is applicable to the device.
+func (run *Runner) Applicable(headers map[string]interface{}) bool {
+	series, err := stringList(headers, "series")
+	if err != nil {
+		return false
+	}
+	if len(series) != 0 && !strutil.ListContains(series, release.Series) {
+		return false
+	}
+	// TODO: architecture, model filtering
+	return true
+}
+
+var errNotToRun = errors.New("repair is not to run")
```

Comment:
Sounds a bit confusing. "repair unnecessary on this system"?  errSkip perhaps, to agree with the status?